src/FastState.cpp

```diff
@@ -62,11 +63,35 @@ void FastState::reset_board() {
 }
 
 bool FastState::is_move_legal(int color, int vertex) const {
-    return vertex == FastBoard::PASS ||
-           vertex == FastBoard::RESIGN ||
-           (vertex != m_komove &&
-                board.get_state(vertex) == FastBoard::EMPTY &&
-                !board.is_suicide(vertex, color));
+    return !is_to_avoid(color, vertex, m_movenum) && (
+              vertex == FastBoard::PASS ||
+                 vertex == FastBoard::RESIGN ||
+                 (vertex != m_komove &&
+                      board.get_state(vertex) == FastBoard::EMPTY &&
+                      !board.is_suicide(vertex, color)));
+}
+
+bool FastState::is_to_avoid(int color, int vertex, size_t movenum) const {
+    for (auto& move : cfg_analyze_tags.m_moves_to_avoid) {
+        if (color == move.color && vertex == move.vertex && movenum <= move.until_move) {
+            return true;
+        }
+    }
+    if (vertex != FastBoard::PASS && vertex != FastBoard::RESIGN) {
+        bool active_allow = false;
+        for (auto& move : cfg_analyze_tags.m_moves_to_allow) {
+            if (color == move.color && movenum <= move.until_move) {
+                active_allow = true;
```

Comment:
Wouldn't we want to set this as soon as cfg_analyze_tags.m_moves_to_allow is not empty? What's the reason it is inside this if?

Reply:
Thanks for the review!

Each `allow` and `avoid` tag has its own move number range.

I.e. `allow b q16 1 avoid b r3 10` will set the `m_moves_to_allow` to a non-empty list. However from move 2 to 10 (and also for all moves beyond that) it as quasi-empty. The `active_allow` flag indicates that we are in a move number range where `allow` is in effect, which means that all non-allowed moves will be avoided.